src/validations/html/rules/MismatchedTag.js

```diff
@@ -0,0 +1,139 @@
+import {List, Map, Record, Set} from 'immutable';
+import Code from '../rules/Code';
+
+const Location = Record({row: 0, column: 0});  // eslint-disable-line new-cap
+
+export default class {
```

Comment:
Provided we only find misplaced closing tags after an open tag, I got a very similar algorithm to work:

- If the closing tag we have encountered matches an opening tag in the stack, add all unclosed tags to collection 1.
- Otherwise:
  - If the closing tag we have encountered matches an entry in collection 1, remove that entry from collection 1, and add the open/close pair to collection 2
  - Otherwise, add the closing tag to collection 3

I think adding all mismatches to collection 1, regardless of whether the mismatch already qualifies for either collection 2 or collection 3, leads to spurious unclosed tag errors since collection 2 or 3 would already "explain" that unclosed tag.